### ISSUE TITLE
Using ND4j updater with SameDiff gradients 

### ISSUE THREAD

**Issue by nadheesh** [deeplearning4j/deeplearning4j#6558]
I'm trying to use ND4j Adam optimizer with a objective that was implemented based on ND4J SameDiff auto differentiation. I don't want to use the `NeuralNetConfiguration.Builder()` or `ComputationGraph` from DeepLearning4j, I only want to use the updater/optimizers with some gradients that is computed with some custom code. 

I found that we can use something similar to below, 

    // list of parameters to be optimized 
    SDVariable[] params; // I have this already initialized. 

    // list of gradients computed for each parameter in params
    INDArray gradients; // I have computed already computed this

    // now I want to use above variables to perform a single gradient descent step based on Adam optimizer
    // based on the source code, we can use something similar to below
    Adam adam = new Adam(0.01);
    GradientUpdater updater = adam.instantiate(INDArray viewArray, boolean initializeViewArray);
    updater.applyUpdater(INDArray gradient, int iteration, int epoch);

But I'm not sure what are the values that we should set to each parameter. Especially, what are the `viewArray` and `initializeViewArray` parameters? 

I appreciate if someone can help me to understand the best way of doing it.
Thanks!!!

**Comment by AlexDBlack** [comment]
SameDiff training (including using the updaters you mentioned) is actually something I'm starting to look at now.

As for the viewArray parameter: that's a row vector (shape [1, updaterStateSize]) array for the updater state - i.e., the history array etc. You can use ```IUpdater.stateSize(long numParams);``` to work out how large the updater state size is (it's usually 0, 1, 2 or 3x the number of parameters that it's used to optimize).

When setting up a new updater, you can pass in a new/uninitialized array for the viewArray parameter, with initializeViewArray=true.
If you are loading a saved updater state from disk (i.e., train -> save -> load -> train more) then use initializeViewArray = false.

**Comment by nadheesh** [comment]
Thanks @AlexDBlack.
If we have several parameters to be optimized, do we have to use a GradientUpdater for each variable since each Updater require storing different viewArrays corresponding to each variable?  

With you help, I was able to implemente a simple linear regression based on only Samediff and NDJ4 (without dl4j). Shall we add that to the examples?

**Comment by nadheesh** [comment]
or can we append all the variables to a single NDArray and then use it with a single updater? I tried to find what you have done in dl4j, yet I could not find that. What is the best and optimal way of handling multiple variables with updaters?

**Comment by AlexDBlack** [comment]
> What is the best and optimal way of handling multiple variables with updaters?

Currently, yes - one updater per variable. Internally, we'll probably optimize this (use fewer updater instances) with our built-in SameDiff training implementation.

> With you help, I was able to implemente a simple linear regression based on only Samediff and NDJ4 (without dl4j). Shall we add that to the examples?

Great. As for adding it to the examples - your call there. We should have full/built-in SameDiff training in the next release (which should be "weeks not months" away) - so as long as you don't mind the example being replaced after the next release, then go ahead and submit a PR :)
